### PR TITLE
Fix drop order of JoinedCell fields

### LINKED ISSUE #20: Incorrect drop order for `JoinedCell`
```rs
use self_cell::self_cell;

struct StrRef<'a>(&'a str);
impl Drop for StrRef<'_> {
    fn drop(&mut self) {
        println!("{}", self.0);
    }
}

self_cell! {
    struct Foo {
        owner: String,
        #[covariant]
        dependent: StrRef,
    }
}

fn main() {
    let foo = Foo::new("Hello World".into(), |s| StrRef(s));
} // prints garbage
```

### PR DESCRIPTION
**Pull request by steffahn**
Closes #20

### CODE REVIEW

**Comment by Voultapher** [comment]
Let me know if you agree with https://github.com/Voultapher/self_cell/pull/21/commits/2ca0660884e55886505633e33f37ea5280f1d8ec

**Comment by Voultapher** [comment]
Also do you think switching to ManuallyDrop would be worth it? We have pointers anyway, so the use seems limited to me. And I suspect this simple pointer based code is quite fast to compile.

**Comment by steffahn** [comment]
2ca0660 Seems like a valid approach as well.

I’m not sure if it’d be superflous, but I’d feel even better if the code used `addr_of_mut!((*ptr).field)` instead of the `&mut (*ptr).field`. I’m just being paranoid that `&mut (*ptr).field` might require that the whole `*ptr` is supposed to be valid, even though a different field is already dropped. (Probably not, but I’m not sure and the change wouldn’t hurt anyway I guess?)

**Comment by steffahn** [comment]
Note that even without 2ca0660, [drop order of fields](https://doc.rust-lang.org/reference/destructors.html) _is_ clearly specified.

But with the `addr_of_mut`, I like it even better than the previous `drop_in_place(joined_ptr.as_ptr());`, because in my mind that one could also try to get mutable access to the whole `*joined_ptr.as_ptr()` first, which would invalidate the shared reference to owner that dependent might contain.

In other words: I’m not quite sure if it makes a difference but if any version is sound then the current one with `drop_in_place(addr_of_mut!((*joined_ptr.as_ptr()).dependent)); drop_in_place(addr_of_mut!((*joined_ptr.as_ptr()).owner));` should be sound, too.

**Comment by steffahn** [comment]
I, too, don’t really think using `ManuallyDrop` adds anything useful.

**Comment by steffahn** [comment]
Squashed the commits a bit.

**Comment by steffahn** [comment]
Added another commit with more changes. I’m more confident that these operations are sound with `addr_of_mut`. In particular, the `UnsafeSelfCell::borrow_mut` seemed problematic because it returned a mutable reference to the whole `JoinedCell` even though `owner` must not be borrowed mutably while `dependent` can hold a reference to it.

**Comment by steffahn** [comment]
Another usage of `addr_of_mut` added. I’m not sure if `&mut (*self.joined_ptr.as_ptr()).owner` is allowed when the `dependent` field contains uninitialized data. Better safe than sorry.

**Comment by steffahn** [comment]
Did some more reading, turns out the `addr_of[_mut]` usage is unnecessary in most cases, but 2 things were legitimate AFAICT:
* the `UnsafeSelfCell::borrow_mut` was problematic, creating an intermediate mutable reference to the whole `JoinedCell`
* the 
  ```rs
  let owner_ptr: *mut $Owner = core::ptr::addr_of_mut!((*joined_ptr.as_ptr()).owner);
  let dependent_ptr: *mut $Dependent = core::ptr::addr_of_mut!((*joined_ptr.as_ptr()).dependent);
  ```
  are important, because previously they would create intermediate references to uninitialized memory.

**Comment by steffahn** [comment]
Reduced usage of `addr_of[_mut]` to those cases now :-), squashed again

File: src/unsafe_self_cell.rs [hunk 1]
```diff
@@ -64,19 +64,29 @@ impl<ContainedIn, Owner, DependentStatic> UnsafeSelfCell<ContainedIn, Owner, Dep
         &(*joined_ptr.as_ptr()).dependent
     }
 
-    pub unsafe fn borrow_mut<'a, Dependent>(&'a mut self) -> &'a mut JoinedCell<Owner, Dependent> {
+    pub unsafe fn borrow_mut<'a, Dependent>(&'a mut self) -> (&'a Owner, &'a mut Dependent) {
```
[the remark this answers is not in the document]
Comment (Voultapher) [reply src/unsafe_self_cell.rs:67]: One could argue it's not part of the documented interface, so we are free to change it. Technically even #22 is a breaking change because one could have written code that observes the change (Hyrum's law).

[the remark this answers is not in the document]
Comment (steffahn) [reply src/unsafe_self_cell.rs:67]: changes to `doc(hidden)` API is usually not considered a breaking change. This doesn’t require a major version bump.
